core/model/modx/processors/search/search.class.php

```diff
@@ -1,57 +1,534 @@
+            $content = $record->get($this->searchField);
+
+            $icon = $this->elements[$this->searchElement]['icon'];
+            $action = $this->elements[$this->searchElement]['action'];
+            $type = $this->elements[$this->searchElement]['type'];
+
+            if (in_array($this->searchField, array('plugincode', 'snippet', 'content', 'introtext'))) {
+                // find line where $search can be found
+                $lines = explode("\n", $content);
+                $content = '';
+                $search = str_replace('%', '', $this->searchQuery);
+
+                $i = 0;
+                foreach ($lines as $linenumber => $line) {
+                    $i++;
+                    if (is_numeric(stripos($line, $search))) {
+                        $line = htmlspecialchars($line);
+                        $linenumber = str_pad($linenumber + 1, 3, '0', STR_PAD_LEFT);
+                        $line = str_replace($search, '<strong>' . $search . '</strong>', $line);
+                        $content .= "<pre class='line'><span class='number'>$linenumber</span><code>$line</code></pre>";
+                    }
                 }
+            } else {
+                $content = str_replace($this->searchQuery, "<strong>{$this->searchQuery}</strong>", $content);
             }
+
+            $results[] = array(
+                'uberbar_header' => $this->modx->lexicon('uberbar_header'),
+                'msg' => '',
+                'options' => '',
+                'header' => '',
+                'id' => $id,
+                'name' => $name,
+                'description' => $description,
+                'content' => $content,
+                'icon' => $icon,
+                '_action' => $action . '&id=' . $id,
+                'type' => $type,
+                'current' => $start++
+            );
         }
+        return $this->outputArray($results);
+    }
 
-        return $this->outputArray($this->results);
+    /**
+     * Collect a list of elements that can be searched, and their token to search by in the uberbar.
+     *
+     * Each element needs to define:
+     * - label
+     * - type
+     * - class_key: the xPDO object class
+     * - permission: boolean if the user has permission for this type of object
+     * - action: where to go in the manager to edit the element
+     * - fields: an array of field names that can be searched in simple mode
+     * - name: the name field to use in the uberbar results
+     * - description: the description field to use in the uberbar results
+     * - content: the field that contains the content, can be shown in the results
+     * - icon: a font awesome icon to use for results of this type
+     */
+    protected function _getElements()
+    {
+        $this->elements['s'] = array(
+            'label' => '(s)nippet',
+            'type' => 'snippets',
+            'class_key' => 'modSnippet',
+            'permission' => $this->modx->hasPermission('view_snippet'),
```

Comment:
Why not populating `$elements[$key]` only if the permission is available, ie. `if ($this->modx->hasPermission('view_snippet')) { $this->elements['s'] = [...] }`. That would prevent iterating over "models" user has no access to